be/src/exec/pipeline/jdbc_scan_operator.cpp

```diff
@@ -0,0 +1,223 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "exec/pipeline/jdbc_scan_operator.h"
+
+#include <sstream>
+
+#include "common/config.h"
+#include "exec/vectorized/jdbc_scanner.h"
+#include "runtime/descriptors.h"
+#include "runtime/jdbc_driver_manager.h"
+#include "util/defer_op.h"
+
+namespace starrocks {
+namespace pipeline {
+
+JDBCScanOperator::JDBCScanOperator(OperatorFactory* factory, int32_t id, int32_t plan_node_id,
+                                   const TJDBCScanNode& jdbc_scan_node, const std::vector<ExprContext*>& conjunct_ctxs,
+                                   int64_t limit)
+        : SourceOperator(factory, id, "jdbc_scan", plan_node_id),
+          _jdbc_scan_node(jdbc_scan_node),
+          _conjunct_ctxs(conjunct_ctxs),
+          _limit(limit),
+          _result_chunks(config::doris_scanner_queue_size) {}
+
+Status JDBCScanOperator::prepare(RuntimeState* state) {
+    SourceOperator::prepare(state);
+    _state = state;
+    _result_tuple_desc = state->desc_tbl().get_tuple_descriptor(_jdbc_scan_node.tuple_id);
+    _start_scanner_thread(state);
+    return Status::OK();
+}
+
+void JDBCScanOperator::close(RuntimeState* state) {
+    if (_scanner_thread) {
+        _scanner_thread->join();
+    }
+    Operator::close(state);
+}
+
+bool JDBCScanOperator::has_output() const {
+    if (_is_finished.load()) {
+        return false;
+    }
+    // some chunks are not being consumed
+    if (!_result_chunks.empty()) {
+        return true;
+    }
+
+    // no new chunks and scanner thread finished successfully
+    Status scanner_status;
+    bool scanner_finished = _is_scanner_finished(&scanner_status);
+    if (scanner_finished && scanner_status.ok()) {
+        return false;
+    }
+    // Here are two situation
+    // 1. scanner thread is still running
+    // 2. scanner thread finished in error, we should make sure `pull_chunk` has a chance to get scanner status
+    return true;
+}
+
+bool JDBCScanOperator::pending_finish() const {
+    return !_is_finished.load();
+}
+
+bool JDBCScanOperator::is_finished() const {
+    Status scanner_status;
+    bool scanner_finished = _is_scanner_finished(&scanner_status);
+    if (scanner_finished) {
+        // scanner thread finished and all chunks have been consumed
+        if (scanner_status.ok() && _result_chunks.empty()) {
+            return true;
+        }
+        // scanner thread finished in error, we should make sure `pull_chunk` has a chance to get scanner status
+        if (!scanner_status.ok()) {
+            return false;
+        }
+    }
+    return _is_finished.load();
+}
+
+Status JDBCScanOperator::_start_scanner_thread(RuntimeState* state) {
+    _scanner_thread.reset(new std::thread(&JDBCScanOperator::_start_scanner, this, state));
+    return Status::OK();
```

Comment:
Maybe use a thread pool for further optimization?

Reply:
> Maybe use a thread pool for further optimization?

of course~

Reply:
for each query we start a new thread may be cause a leak in thread_local variables. we need add a TODO to test it later.

Reply:
> for each query we start a new thread may be cause a leak in thread_local variables. we need add a TODO to test it later.

@stdpain fixed